go/types/sequence_cursor.go

```diff
@@ -32,16 +62,115 @@ func (cur *sequenceCursor) getItem(idx int) sequenceItem {
 	return cur.seq.getItem(idx)
 }
 
+/*
+ * sync advances the cursor to the next chunk
+ */
 func (cur *sequenceCursor) sync() {
 	d.PanicIfFalse(cur.parent != nil)
 	cur.seq = cur.parent.getChildSequence()
+	cur.readAheadCh = nil
+}
+
+/*
+ * doReadAhead returns true if read-head should be used when reading leaves
+ * of this cursor.
+ *
+ * Only read-ahead if the current sequence is a meta-sequence that contains
+ * leaf sequences.
+ */
+func (cur *sequenceCursor) doReadAhead() bool {
+	if cur.readAhead && cur.readAheadCh == nil {
+		if ms, ok := cur.seq.(metaSequence); ok {
+			return ms.tuples[0].ref.height == 1
+		}
+	}
+	return false
+}
+
+// Records in read-ahead channel
+type readAheadRec struct {
+	idx     int
+	leafSeq sequence
+}
+
+var readAheadParallelism = runtime.NumCPU() * 8
+
+/*
+ * readAheadLeaves initiates read-ahead of child sequences.
+ *
+ * Fills a buffered read-ahead channel with child sequences in ascending index order.
+ * A forward moving cursor reads this channel to find it's next item (see getChildSequence).
+ *
+ * The channel is limited to c * NumCPU. This allows read-ahead to proceed in parallel
+ * while maintaining a bound on memory.
+ */
+func (cur *sequenceCursor) readAheadLeaves() {
+	ms := cur.seq.(metaSequence)
+	startIdx := cur.idx
+	tuplesToRead := len(ms.tuples) - startIdx
+	if tuplesToRead < 2 {
+		// Don't read-ahead if we're starting at the end of a sequence. If the cursor's moving
+		// backward, read-ahead is useless. If the cursor's moving forward, there's no
+		// benefit.
+		cur.readAheadCh = nil
+	} else {
+		parallelism := int(math.Min(float64(readAheadParallelism), float64(tuplesToRead)))
+		input := make(chan interface{}, parallelism)
+
+		cur.readAheadCh = orderedparallel.New(input, func(item interface{}) interface{} {
+			i := item.(int)
+			return &readAheadRec{i, ms.getChildSequence(i)}
+		}, parallelism)
+
+		go func() {
```

Comment:
I think the biggest problem here is that you have no way to stop this goroutine if the caller decides to only iterate part of the sequence. That's a problem because it means the buffered channel and this goroutine will stick around forever. Depending on what's actually captured in the closure above, the cursor could wind up living forever as well.

Now, I don't think this API has any way to indicate that the caller is done with the cursor either, so that might be something we need to add :-/ I'd want @kalman to weigh in on that